From 7fbc87976b4fdfc411cb205c26cdb6e56a3202ac Mon Sep 17 00:00:00 2001
From: Jagpal Singh Gill <paligill@gmail.com>
Date: Wed, 14 May 2025 00:55:59 -0700
Subject: [PATCH 2/2] add events and errors for firmware update

Add various events and errors messages that can be generated during the
firmware update process. These event definitions are based on [1], [2]
and [3].

[1]: https://redfish.dmtf.org/registries/Update.1.1.1.json
[2]: https://redfish.dmtf.org/registries/Base.1.21.0.json
[2]: https://www.dmtf.org/sites/default/files/standards/documents/DSP2062_1.0.2.pdf

Tested: meson build passed.

Change-Id: I00e64a9b9f712ae336f49e2cc50da413b65d5b04
Signed-off-by: Jagpal Singh Gill <paligill@gmail.com>
---
 .../Software/Update/meson.build               |   5 +
 gen/xyz/openbmc_project/Software/meson.build  |  23 ++++
 .../Software/Update.events.yaml               | 101 ++++++++++++++++++
 3 files changed, 129 insertions(+)
 create mode 100644 yaml/xyz/openbmc_project/Software/Update.events.yaml

diff --git a/gen/xyz/openbmc_project/Software/Update/meson.build b/gen/xyz/openbmc_project/Software/Update/meson.build
index 5d195c0..a20ad01 100644
--- a/gen/xyz/openbmc_project/Software/Update/meson.build
+++ b/gen/xyz/openbmc_project/Software/Update/meson.build
@@ -6,11 +6,14 @@ generated_sources += custom_target(
     'xyz/openbmc_project/Software/Update__cpp'.underscorify(),
     input: [
         '../../../../../yaml/xyz/openbmc_project/Software/Update.errors.yaml',
+        '../../../../../yaml/xyz/openbmc_project/Software/Update.events.yaml',
         '../../../../../yaml/xyz/openbmc_project/Software/Update.interface.yaml',
     ],
     output: [
         'error.cpp',
         'error.hpp',
+        'event.cpp',
+        'event.hpp',
         'common.hpp',
         'server.hpp',
         'server.cpp',
@@ -32,6 +35,8 @@ generated_sources += custom_target(
     ],
     install: should_generate_cpp,
     install_dir: [
+        false,
+        get_option('includedir') / sdbusplus_current_path,
         false,
         get_option('includedir') / sdbusplus_current_path,
         get_option('includedir') / sdbusplus_current_path,
diff --git a/gen/xyz/openbmc_project/Software/meson.build b/gen/xyz/openbmc_project/Software/meson.build
index 1237155..f959fee 100644
--- a/gen/xyz/openbmc_project/Software/meson.build
+++ b/gen/xyz/openbmc_project/Software/meson.build
@@ -331,6 +331,7 @@ generated_markdown += custom_target(
     'xyz/openbmc_project/Software/Update__markdown'.underscorify(),
     input: [
         '../../../../yaml/xyz/openbmc_project/Software/Update.errors.yaml',
+        '../../../../yaml/xyz/openbmc_project/Software/Update.events.yaml',
         '../../../../yaml/xyz/openbmc_project/Software/Update.interface.yaml',
     ],
     output: ['Update.md'],
@@ -352,6 +353,28 @@ generated_markdown += custom_target(
     build_by_default: should_generate_markdown,
 )
 
+generated_registry += custom_target(
+    'xyz/openbmc_project/Software/Update__registry'.underscorify(),
+    input: ['../../../../yaml/xyz/openbmc_project/Software/Update.events.yaml'],
+    output: ['Update.json'],
+    depend_files: sdbusplusplus_depfiles,
+    command: [
+        sdbuspp_gen_meson_prog,
+        '--command',
+        'registry',
+        '--output',
+        meson.current_build_dir(),
+        '--tool',
+        sdbusplusplus_prog,
+        '--directory',
+        meson.current_source_dir() / '../../../../yaml',
+        'xyz/openbmc_project/Software/Update',
+    ],
+    install: should_generate_registry,
+    install_dir: [inst_registry_dir / sdbusplus_current_path],
+    build_by_default: should_generate_registry,
+)
+
 generated_markdown += custom_target(
     'xyz/openbmc_project/Software/Version__markdown'.underscorify(),
     input: [
diff --git a/yaml/xyz/openbmc_project/Software/Update.events.yaml b/yaml/xyz/openbmc_project/Software/Update.events.yaml
new file mode 100644
index 0000000..b7e9a05
--- /dev/null
+++ b/yaml/xyz/openbmc_project/Software/Update.events.yaml
@@ -0,0 +1,101 @@
+version: 1.0.0
+
+errors:
+    - name: VerificationFailed
+      severity: critical
+      metadata:
+          - name: ImageIdentifier
+            type: string
+            primary: true
+            description:
+                The identifier for the image. This shall map to
+                PackageVersionString for a PLDM package and version field for a
+                manifest based image format.
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+      errno: EINVAL
+      redfish-mapping: Update.VerificationFailed
+
+    - name: ApplyFailed
+      severity: critical
+      metadata:
+          - name: ImageIdentifier
+            type: string
+            primary: true
+            description:
+                The identifier for the image. This shall map to
+                PackageVersionString for a PLDM package and version field for a
+                manifest based image format.
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+      errno: EFAULT
+      redfish-mapping: Update.ApplyFailed
+
+events:
+    - name: TargetDetermined
+      metadata:
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+          - name: ImageIdentifier
+            type: string
+            primary: true
+            description:
+                The identifier for the image. This shall map to
+                PackageVersionString for a PLDM package and version field for a
+                manifest based image format.
+      redfish-mapping: Update.TargetDetermined
+
+    - name: UpdateInProgress
+      metadata:
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+          - name: ProgressPercentage
+            type: size
+            primary: true
+            description: The amount of progress achieved in percentage.
+      en:
+          description: The update is in progress for related target.
+          message:
+              An update is in progress for {TargetName} with
+              {ProgressPercentage} percent complete.
+
+    - name: UpdateSuccessful
+      metadata:
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+          - name: ImageIdentifier
+            type: string
+            primary: true
+            description:
+                The identifier for the image. This shall map to
+                PackageVersionString for a PLDM package and version field for a
+                manifest based image format.
+      redfish-mapping: Update.UpdateSuccessful
+
+    - name: ResetDevice
+      metadata:
+          - name: TargetName
+            type: object_path
+            primary: true
+            description:
+                The name or identifier of the firmware target or resource.
+          - name: ResetType
+            type: enum[xyz.openbmc_project.State.Host.Transition]
+            primary: true
+            description: The type of reset operation to perform.
+      redfish-mapping: Base.ResetRequired
-- 
2.47.3

